modyn/evaluator/internal/pytorch_evaluator.py

```diff
@@ -0,0 +1,184 @@
+
+        self._info("Initialized PyTorch evaluator.")
+
+    def _prepare_dataloader(self, evaluation_info: EvaluationInfo) -> torch.utils.data.DataLoader:
+        self._debug("Creating EvaluationDataset.")
+        dataset = EvaluationDataset(
+            evaluation_info.dataset_id,
+            evaluation_info.bytes_parser,
+            evaluation_info.transform_list,
+            evaluation_info.storage_address,
+            evaluation_info.evaluation_id,
+        )
+        self._debug("Creating DataLoader.")
+        dataloader = torch.utils.data.DataLoader(
+            dataset, batch_size=evaluation_info.batch_size, num_workers=evaluation_info.num_dataloaders
+        )
+
+        return dataloader
+
+    def _info(self, msg: str) -> None:
+        self.logger.info(f"[Evaluation {self._evaluation_id}] {msg}")
+
+    def _debug(self, msg: str) -> None:
+        self.logger.debug(f"[Evaluation {self._evaluation_id}] {msg}")
+
+    def _load_state(self, path: pathlib.Path) -> None:
+        assert path.exists(), "Cannot load state from non-existing file"
+
+        self._info(f"Loading model state from {path}")
+        with open(path, "rb") as state_file:
+            checkpoint = torch.load(io.BytesIO(state_file.read()))
+
+        assert "model" in checkpoint
+        self._model.model.load_state_dict(checkpoint["model"])
+
+        # delete trained model from disk
+        os.remove(path)
+
+    def send_status_to_server(self, batch_number: int) -> None:
+        self._status_response_queue.put({"num_batches": batch_number, "num_samples": self._num_samples})
+
+    def evaluate(self) -> None:
+        self._info(f"Process {os.getpid()} starts evaluation.")
+
+        y_true = []
+        y_score = []
+
+        self._model.model.eval()
+        with torch.no_grad():
+            batch_number = -1
+            for batch_number, batch in enumerate(self._dataloader):
+                # As empty() is unreliable
+                # we try to fetch an element within 10ms. If there is no
+                # element within that timeframe returned, we continue.
+                try:
+                    req = self._status_query_queue.get(timeout=0.01)
+                    if req == EvaluatorMessages.STATUS_QUERY_MESSAGE:
+                        self.send_status_to_server(batch_number)
+                    else:
+                        raise ValueError("Unknown message in the status query queue")
+                except queue.Empty:
+                    pass
+
+                data: Union[torch.Tensor, dict]
+                if isinstance(batch[1], torch.Tensor):
+                    data = batch[1].to(self._device)
+                elif isinstance(batch[1], dict):
+                    data: dict[str, torch.Tensor] = {}  # type: ignore[no-redef]
+                    for name, tensor in batch[1].items():
+                        data[name] = tensor.to(self._device)
```

Comment:
Could you open an issue that after merging #256 and this PR, we should also try to extract common code of the trainer and the evaluator into some helper class? to reduce code duplication, basically

Reply:
Done, see #284

Reply:
thanks!